torchvision/datasets/_stereo_matching.py

```diff
@@ -0,0 +1,1041 @@
+
+        self._images: List[Tuple[str, str]] = []
+        self._disparities: List[Tuple[str, str]] = []
+
+    def _read_img(self, file_path: str) -> Image.Image:
+        img = Image.open(file_path)
+        if img.mode != "RGB":
+            img = img.convert("RGB")
+        return img
+
+    def _scan_pairs(
+        self, paths_left_pattern: str, paths_right_pattern: str, fill_empty: bool = False
+    ) -> List[Tuple[str, str]]:
+        left_paths: List[str] = sorted(glob(paths_left_pattern))
+        right_paths: List[str] = sorted(glob(paths_right_pattern))
+
+        # used when dealing with inexistent disparity for the right image
+        if fill_empty:
+            right_paths = list("" for _ in left_paths)
+
+        if not left_paths:
+            raise FileNotFoundError(f"Could not find any files matching the patterns: {paths_left_pattern}")
+
+        if not right_paths:
+            raise FileNotFoundError(f"Could not find any files matching the patterns: {paths_right_pattern}")
+
+        if len(left_paths) != len(right_paths):
+            raise ValueError(
+                f"Found {len(left_paths)} left files but {len(right_paths)} right files using:\n "
+                f"left pattern: {paths_left_pattern}\n"
+                f"right pattern: {paths_right_pattern}\n"
+            )
+
+        images = list((left, right) for left, right in zip(left_paths, right_paths))
+        return images
+
+    @abstractmethod
+    def _read_disparity(self, file_path: str) -> Tuple:
+        # function that returns a disparity map and an occlusion map
+        pass
+
+    def __getitem__(self, index: int) -> Tuple:
+        """Return example at given index.
+
+        Args:
+            index(int): The index of the example to retrieve
+
+        Returns:
+            tuple: A 4-tuple with ``(img_left, img_right, disparity, valid_mask)`` where ``valid_mask``
+                is a numpy boolean mask of shape (H, W)
+                indicating which disparity values are valid. The disparity is a numpy array of
+                shape (1, H, W) and the images are PIL images. ``disparity`` and ``valid_mask`` are None for
+                datasets on which for ``split="test"`` the authors did not provide annotations.
+        """
+        img_left = self._read_img(self._images[index][0])
+        img_right = self._read_img(self._images[index][1])
+
+        dsp_map_left, valid_mask_left = self._read_disparity(self._disparities[index][0])
+        dsp_map_right, valid_mask_right = self._read_disparity(self._disparities[index][1])
+
+        imgs = (img_left, img_right)
+        dsp_maps = (dsp_map_left, dsp_map_right)
+        valid_masks = (valid_mask_left, valid_mask_right)
+
+        if self.transforms is not None:
+            (
+                imgs,
+                dsp_maps,
+                valid_masks,
+            ) = self.transforms(imgs, dsp_maps, valid_masks)
+
+        return imgs[0], imgs[1], dsp_maps[0], valid_masks[0]
```

Comment:
I believe it would just streamline transform definition / writing for the end user. Input / Output is isometric, they just have to handle how that input is being processed inside the transform.

Reply:
For this, I also prefer for the dataset to always output same number of output (always has valid_mask although it can be None). It feel less "unexpected" to me.

One design question I have for valid_mask is when the dataset dont provide any file on valid_mask, should we generate the valid_mask on `_read_disparity_files` or should we just leave it None? (in this case we let the transform to define the valid_mask)

Reply:
The proposal I've included in the PR is to basically make a mask full of ones when we don't get any information regarding that. Since the usage of the mask is to indicate on which pixels in the disparity map we should compute the loss a `None` would be equivalent to evaluating the loss on the entire outputs.

Reply:
> One design question I have for valid_mask is when the dataset dont provide any file on valid_mask

I think we should let the transform define it, like we do for OF datasets.

I agree that we should try to let the dataset always return the same number of outputs. This is why we let `flow` or `disparity` be `None` in "test" splits, where it's not available, as it's only available in "train".

However, in the case of `valid_mask` which may not always exist, it doesn't seem expected to always return it if it is not defined by the dataset. Taken in isolation, why should a dataset return something it doesn't actually have? This was the reason we decided not to return the `valid_flow_masks` in datasets that don't provide them (again, I'm not claiming this is a perfect solution).

As pointed out by @TeodorPoncu , it does make the writing of the transforms a tiny bit trickier. But this is why we provide the presets in our training recipes, which are able to handle all of this for the user, so this is mostly transparent and they don't need to write `if/else-y` code.

I would still err on the side of consistency with the rest of our API for this, as this is one of the most important aspect of a frictionless user experience

Reply:
Speaking from anecdotal experience most of the people that I know who use datasets from torchvision don't even look at the training reference scripts / know that they exists.

I don't have any strong feelings towards moving the valid mask into the transform part, however I feel most users would end up writing their own logic to handle various shape returns.

I think there's multiple ways of thinking about it: Dataset centric way or Task centric way. In the dataset centric way you'd want to get all the available data that the dataset contains and the user would be responsible for selecting reshaping processing w/e the outputs for his task.

In the Task centric way as a user you'd want to always get the necessary inputs / outputs such that you can perform the task (irrespective if the dataset has them or not). If the dataset does not contain that specific piece of information, you'd want to get some sort of output for it which tells you that information does not exist or that it gives you a mathematical equivalent for the task. (in this case a mask full of ones)

Reply:
To me personally, transforms imply modifying / altering data, not necessarily creating or extracting information and as a personal preference I'd avoid using transforms that transform data types / generate data.

Another argument that I have for consistent shape outputs from __getitem__ is such that you can guarantee transform associativity as often as possible.

Reply:
Thanks @TeodorPoncu , these are all good points TBH.

> Speaking from anecdotal experience most of the people that I know who use datasets from torchvision don't even look at the training reference scripts / know that they exist.

Ah, that's a shame. Perhaps this is because we didn't do a great job at documenting these in the past. Hopefully the new doc revamp will help raising awareness about them.

> Dataset centric way or Task centric way

I agree with your analysis. For OF datasets we opted for the dataset-centric way. I don't think that there is a clear winner **a-priori**, but now that we already went this way for OF datasets, honouring consistency is critical.

> To me personally, transforms imply modifying / altering data, not necessarily creating or extracting information and as a personal preference I'd avoid using transforms that transform data types / generate information.

I tend to agree as well. We still decided to generate the fake data in the transforms rather in the datasets, because generating it in the datasets has various drawbacks:

- it bounds to user to *one* specific generation, which may not be the one they want. For example in optical flow we generate a fake `valid_mask` depending on a given threshold, but the default threshold may not suit all users:

  https://github.com/pytorch/vision/blob/1dd17538d98f960bcdfa35cab7fd51efbc73cbbf/references/optical_flow/transforms.py#L33:L33

   (we could make this a parameter of the dataset class, but it's not ideal either)

- it  generates a lot of useless data and thus uses memory, while some users may not actually need the `valid_mask` at all. So it's best to avoid generating it on their behalf.

> Another argument that I have for consistent shape outputs from getitem is such that you can guarantee transform associativity as often as possible.

I agree this is important, and we can have that regardless of the design :) In the OF datasets, we don't return (or generate fake) values that don't exist, but we still require all the transforms to have the same signature: `img1, img2, flow, valid_mask`:
-  https://github.com/pytorch/vision/blob/add-crestereo-dataset/torchvision/datasets/_optical_flow.py#L65:L65
- https://github.com/pytorch/vision/blob/1dd17538d98f960bcdfa35cab7fd51efbc73cbbf/torchvision/datasets/_optical_flow.py#L242-L245

Reply:
I believe all of these concerns are covered by the existing design we have for optical flow datasets. I provided references above but I'm happy to provide more details offline/in person if you wish.

Reply:
No, it's not necessary. I do agree that datasets should remain pure and we should not make any assumptions regarding masks. My final question would be w.r.t. to datasets (such as Kitti2015) for which some authors (i.e. [Raft-STEREO](https://github.com/princeton-vl/RAFT-Stereo/blob/0e2a12746143a7552e30ef2f4b1d4c3214388a1a/core/utils/frame_utils.py#L124)) have some basic methods of creating a mask.

If paper authors / dataset authors provide a method for computing a valid mask method (i.e [CREStereo](https://github.com/megvii-research/CREStereo/blob/ad3a1613bdedd88b93247e5f002cb7c80799762d/dataset.py#L91)) should we consider that a valid way of generating masks and add it into our dataset implementations or if there's no occlusion / valid map file should we just not return anything at all?